Automattic/wp-calypso
# Improve calypso app builds

### Changes proposed in this Pull Request
Each calypso app has something in common: they are deployed to a location on WordPress.com. Thus far, we have only really supported syncing to WordPress.com for ETK, with support for the other apps often being broken or nonexistent.

The goal of this PR is to make each app behave the same way:
1. `yarn dev` will run webpack, watching for changes.
2. `yarn build` will run webpack in production mode without watching for changes.
3. Adding `--sync` to either of the above commands will sync the changes to the correct location on WordPress.com


#### Unified build script

We accomplish this with a bin script which launches the build commands. Each package gets the following commands:
```json
{
	"build:some-entry": "calypso-build --source='some/entrypoint'",
	"clean": "npx rimraf dist",
	"dev": "node ../../bin/build-calypso-app.mjs --localPath dist --remotePath /home/wpcom/public_html/path",
	"build": "NODE_ENV=production yarn dev",
}
```

An app may have several entrypoints, each specified under a `build:` script. These are all run in parallel by `npm-run-all`. The user can manually specify `--sync`, and this will be passed to the build-calypso-app script.

The builder script first cleans the app, then runs all `build:*` tasks in parallel. It then conditionally syncs to the remote depending on whether `--sync` exists. It syncs once in production mode, and listens for syncs in development mode. The sync command itself is debounced to 500ms to avoid overwhelming the connection while webpack is building.

This approach allows us to simplify the ETK package.json, and it allows us to share the same build script across all calypso apps. This way, a fix to the build script will apply everywhere.

#### Fixes for app webpack builds:
1. Resolves the ETK issue where Node says there is a memory leak because there are too many event emitters. By increasing the event emitter limit to 20, we avoid the problem.
2. Resolves the issue with webpack 5 deprecation warnings. These were coming from the duplicate package plugin, which is already disabled for calypso client builds. I've disabled that by default for all calypso apps. 

#### Output changes:
1. A few ETK files register as "changed". This is because I unified a few `build:` scripts into the same one. `calypso-build` can accept multiple sources, so we didn't need to split these scripts out into separate build tasks. We only need separate build tasks for when the dist folder is separate, which was not the case for the ones I changed.

### Todo:
- [x] Process failure does not close child processes. For example, if rsync fails, the script will exit, but webpack processes will be left dangling. 
- [x] rsync processes may not be getting terminated

### Testing instructions
for each of the wpcom apps:
1. Run `yarn build` and `yarn dev` locally and verify they work properly.
2. Run `yarn dev --sync` and test that syncing to wpcom and watching for changes works properly.
3. Verify that the webpack bundles in TeamCity have not changed in any meaningful way.
